### PR TITLE
Frontend: friends list with React Query

### PR DESCRIPTION
**Pull request by darrenvong**
Lays out the structure for #291 with the initial friend list code implemented with React Query.

This feature is surprisingly complex enough to allow me to play around React Query and demo a few things:
- Dependent queries (i.e. we only fetch the required users after we find out who the user's friend (ID)s are)
- Dynamic number of queries with the new `useQueries` (plural) hook. Dynamic as it depends how many friends a user has!
- Persisting the cache in local storage with the experimental `react-query/persist-localstorage-experimental` package built into React Query
- Devtools from React Query in dev mode that lets you see what's going on - i.e. when React Query refetches and marks a query as stale etc.

There are lots of options to tweak if certain defaults are too "aggressive", but I find them to be working ok so far in this demo, that means it's zero config as promised in the docs so far!

### CODE REVIEW

File: app/frontend/src/features/connections/friends/FriendList.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/FriendList.tsx]: Can we avoid `as`?

File: app/frontend/src/features/connections/friends/useFriendList.ts
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.ts]: Can this be simplified to something like

```
const errors = [error?.message , ... friendQueries.map((q) => q.error?.message).where((e) => e !== null);
```

**Review by lucaslcode — COMMENTED** [review]
Think this looks really nice, let's discuss more in the meeting

File: app/frontend/src/features/connections/friends/useFriendList.ts
Comment (lucaslcode) [reply app/frontend/src/features/connections/friends/useFriendList.ts]: Across multiple lines ofc, I'm on my phone

File: app/frontend/src/features/connections/friends/FriendList.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/FriendList.tsx]: Sadly not in this case, since `useQueries` is a generic function that accepts any type of queries, React Query doesn't provide a good way to type it and types it as `unknown` under the hood. When I try to type `payload`, I keep on getting variants of `unknown` is not assignable to `User.AsObject` ...

Comment (lucaslcode) [reply app/frontend/src/features/connections/friends/FriendList.tsx]: I see, how about moving the type assertion into the hook instead? Only if it's easy.

In an closed stale issue they mention they might accept a PR they puts generic types in, I might try that. Could be straightforward.

Comment (darrenvong) [reply app/frontend/src/features/connections/friends/FriendList.tsx]: Aha it's a bit of a hack, as if they ever update the signature, we'd have to update the types ourselves as well, but I managed to augment/extend the `react-query` types to allow `useQueries` to be typed with generics, so no more `as` casting/type assertions!

**Comment by darrenvong** [comment]
Closing since this will be replaced by the React Query implementation in the demo PR as agreed in the last meeting.

**Comment by darrenvong** [comment]
Finally got around to wrapping this up with some tests, ready to be reviewed again

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (darrenvong) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:13]: Inlining this in the wrapper is key to make sure a new client/cache is created for each test so they are isolated. This was causing me so much headaches before I realised!

File: app/frontend/src/service/__mocks__/index.ts
Comment (darrenvong) [inline app/frontend/src/service/__mocks__/index.ts:1]: Deleted this as this was also causing a lot of headache when trying to debug, since console logging a proxy gives you back a Symbol object...

Jest mocking correctly mocks out everything deeply anyway when we do `jest.mock` on this, so this is unnecessary code and it's interfering with the good default. And since the implementations in each mocks get reset, this is only good for one test anyway so it's useless. I've moved the default implementations into its own file in the `test` folder where all the test helpers are instead.

File: app/frontend/src/App.css
Comment (lucaslcode) [inline app/frontend/src/App.css]: Why this? Not sure we should start using this file.

File: app/frontend/src/features/connections/friends/FriendList.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/FriendList.tsx]: Is this style specific to the connections `CircularProgress`?

File: app/frontend/src/features/connections/ConnectionsPage.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/ConnectionsPage.tsx]: Is there an easy way to make tabs "routes"? Maybe a `<Switch>` instead of the tab panels? Would probably need a `useEffect` to set `value` too. This is also the case for the messages page, later.

File: app/frontend/src/features/connections/friends/FriendList.stories.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/FriendList.stories.tsx]: Is there a store needed in here?

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:49]: Why is waiting necessary after the assertions?

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:61]: Is this because the default test user has 2 friends? If so maybe put a comment.

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:152]: I am going to steal these for `useUsers` tests :D

File: app/frontend/src/features/connections/friends/useFriendList.ts
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.ts]: I think we should be returning the data rather than the queries. The info on the queries is covered by isLoading, isError and errors. Then it is more similar to `useQuery`

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:93]: This isn't passing

File: app/frontend/src/service/__mocks__/index.ts
Comment (lucaslcode) [reply app/frontend/src/service/__mocks__/index.ts:1]: Sure. How are most of the tests still passing without changes? Because they didn't really use this?

**Review by lucaslcode — CHANGES_REQUESTED** [review]
Nice work, only some relatively minor stuff

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.test.tsx:49]: This was there to get rid of that warning which says there's state updating outside of `act`, even though we don't care it about it in this case since we are interested in the loading state in this test rather than the eventual state when things get resolved successfully.

File: app/frontend/src/service/__mocks__/index.ts
Comment (darrenvong) [reply app/frontend/src/service/__mocks__/index.ts:1]: Yeah this wasn't being used, the existing tests either set the mock implementation before each test or had to manually override the default mock implementation to trigger certain scenarios (e.g. throwing an error).

File: app/frontend/src/App.css
Comment (darrenvong) [reply app/frontend/src/App.css]: This was actually for Storybook as the layout was using `content-box` for some strange reasons. I'll look at moving this into a component on the global Storybook decorator instead.

File: app/frontend/src/features/connections/friends/FriendList.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/FriendList.tsx]: Yeah, it's specifically for centering it to the middle of the tile/container

File: app/frontend/src/features/connections/ConnectionsPage.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/ConnectionsPage.tsx]: This was copied/adapted from the messages stuff which looks not too different from the example code in the Material UI page, so not sure without playing around with it a bit more tbh.

I did think that was annoying for messages though as currently there's no way to copy/paste a link that would get you straight to the group chat, but that might be for a separate PR/issue.

File: app/frontend/src/features/connections/friends/useFriendList.ts
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.ts]: Good point. Although that got me thinking, and I haven't done it at the moment, but do we want to ever display errors at a granular level? For example, some friends loaded successfully but 1-2 errored for some reason. 

Or do we only display their friends when everyone has successfully loaded, otherwise we display an error over the panel/tile?

Comment (lucaslcode) [reply app/frontend/src/features/connections/friends/useFriendList.ts]: I think in this case it's not necessary.

Hmm... Unless the error is there's no connection. What happens in that case and there are still cached users? `isError` is true but data is defined? That is something I need to think about for `useUsers` too.

Comment (lucaslcode) [reply app/frontend/src/features/connections/friends/useFriendList.ts]: If the returned data is a `Map<number, User.AsObject|undefoned>` we could display unloaded users as "(error loading user)" as well as the error message. 

Anyway do what you think is best/easiest for now. We will probably change this to use `useUsers` once that's merged and we can think about it then.

Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.ts]: Yeah `isError` will be true with data defined in case of cached user, which is a bit of a weird state to be in. I imagine in that case we'd want to keep displaying the user?

Only brought this up because of how I defined `isError` here too. I went with the optimistic definition of only saying there is an error if everything errored, but that definitely complicate things.

A blanket error might be the easiest approach for now (although might not be best UX?) and like you said we can re-visit later.

File: app/frontend/src/features/connections/ConnectionsPage.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/ConnectionsPage.tsx]: Addressed this for the friend list tab by using URL params to manage state. The cool thing about that is that if the user visits a specific URL, it will jump straight to that tab rather than always defaulting to the first tab.

I think I'll leave the message tabs as a future fix.

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx]: Was this necessary to not fail? I'd rather avoid waiting in tests. Not sure if that's possible though.

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx]:
```suggestion
        : new Promise(() => null);
```

Would that make a never-resolving promise, or would it just throw?

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (lucaslcode) [inline app/frontend/src/features/connections/friends/useFriendList.test.tsx:207]: Hmm this means if the user's connection drops and they have 20 friends and they try to load the friends list, they will get an alert box with 21 lines. What do you think? Should we just have one "error" per hook and print individual errors to the console with `console.error()`?

Comment (lucaslcode) [reply app/frontend/src/features/connections/friends/useFriendList.test.tsx:207]: Could save for another pr if you want.

**Review by lucaslcode — DISMISSED** [review]
lgtm, feel free to ignore comments

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.test.tsx]: Yeah that wait can't be avoided, since any promise based ops (even ones that have already resolved) won't get its result until the next tick of the event loop, so the wait lets this run for one tick. It's not really a problem if the promise will resolve straight away - most of these tests take like ms to run.

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.test.tsx]: I thought this might cause some weird timeout from jest but it actually doesn't, think I'll change it to this when I get some time later so I don't need to specify a time for `waitForNextUpdate` to wait. It seems to know to wait for only resolved promises and not pending ones so will make the test a bit clearer than randomly waiting for 10ms as a hack.

File: app/frontend/src/features/connections/friends/useFriendList.test.tsx
Comment (darrenvong) [reply app/frontend/src/features/connections/friends/useFriendList.test.tsx:207]: Yeah, I think I'll leave this for when we might refactor this hook to use your `useUsers` hook

**Comment by darrenvong** [comment]
Sorry, made the small tweaks you suggested so it dismissed the approval 😅

**Review by lucaslcode — APPROVED** [review]

**Pull request merged**
